LE-66: A new ORA2 student responses report with emails addresses.

lydiali opened the pull request:
The ORA2 student responses report download (accessible from instructor dashboard -> "data download")
doesn't include the students' email addresses or usernames. We've had requests to include this.

It'd be ideal if there could be a way to request it both with and without the email addresses / usernames.
One straightforward way to present that option would be by simply creating an additional button for
the report including email addresses, and relabelling the existing button.

lydiali commented:
@dylanrhodes @azimpradhan 
Please review.

dcadams commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -2133,6 +2133,32 @@ def get_course_forums_usage(request, course_id):
+    """
+    Pushes a Celery task which will aggregate ora2 responses for a course into a .csv
+    """
+    course_key = SlashSeparatedCourseKey.from_deprecated_string(course_id)
```
As per the discussion about opaque keys we shouldn't be using from_deprecated_string now. Here's a link to the doc:
https://github.com/edx/edx-platform/wiki/Locators%3A-Developer-Notes

lydiali replied:
@dcadams 
According to the link, it says to use this:
     course_key = SlashSeparatedCourseKey.from_string(course_id)
But from_string also seems to be deprecated.  

Should I be using this instead?  
    course_key = locator.CourseLocator.from_string(course_id)

dcadams replied:
Yes, this is something we're all trying to figure out right now. I believe the relevant sections are:

To serialize OpaqueKeys into a string representation, call foo.to_string(), where foo is some type of OpaqueKey object (e.g., either CourseKey, UsageKey, CourseLocator, or BlockUsageLocator).

But SlashSeparatedCourseKey will only work with old style course_ids.

So, I looked for guidance in the upstream master codebase and didn't have to look that far (courseware.views) to find this:

course_key = SlashSeparatedCourseKey.from_deprecated_string(course_id)

dylanrhodes commented on lms/djangoapps/instructor_task/api.py:
```diff
@@ -23,7 +23,9 @@
     get_ora2_responses,
+    get_ora2_responses_email,
     get_course_forums_usage,
+
```
Nit: remove this empty line, please

dylanrhodes commented on lms/djangoapps/instructor_task/tasks.py:
```diff
@@ -32,6 +32,8 @@
     push_student_responses_to_s3,
     push_ora2_responses_to_s3,
+    push_ora2_responses_to_s3_email,
+    push_student_forums_data_to_s3,
```
It seems like this import shouldn't be a part of this PR, but rather Jim's or Azim's.  Does this PR touch the student forums data report?

dylanrhodes commented on lms/djangoapps/instructor_task/tasks.py:
```diff
@@ -199,3 +202,14 @@ def get_student_forums_usage(entry_id, xmodule_instance_args):
+@task(base=BaseInstructorTask, routing_key=settings.ORA2_RESPONSES_DOWNLOAD_ROUTING_KEY)
+def get_ora2_responses_email(entry_id, xmodule_instance_args):
+    """
+    Generate a CSV of ora2 responses and push it to S3.
```
It would be nice to mention that it includes emails in the comment to disambiguate it from the existing get_ora2_responses function.

dylanrhodes replied:
If you take my advice above and go for a parameterized endpoint instead, though, we won't need this duplicate method.

dylanrhodes commented on lms/djangoapps/instructor_task/tasks_helper.py:
```diff
@@ -720,6 +724,7 @@ def update_task_progress():
         filename,
         course_id,
         start_time,
+
```
Nit: I don't think we want this empty line.

dylanrhodes commented on lms/djangoapps/instructor_task/tasks_helper.py:
```diff
@@ -647,11 +646,11 @@ def push_student_responses_to_s3(_xmodule_instance_args, _entry_id, course_id, _
 
 
-def push_ora2_responses_to_s3(_xmodule_instance_args, _entry_id, course_id, _task_input, action_name):
+def push_ora2_responses_to_s3_base(_xmodule_instance_args, filename, _entry_id, course_id, _task_input, action_name, is_anonymous):
```
I think it makes more sense for the push_ora2_responses_to_s3 and push_ora2_emails_to_s3 methods to call _push_csv_responses_to_s3 directly rather than through this wrapper.  It will make reading the code easier in the future, since they're currently located in different sections of the file.  It also seems like overkill to have a single-line wrapper function, even if it does abstract a little bit of work from the functions calling it.

azimpradhan replied:
Yes, +1 on this

dylanrhodes commented on lms/djangoapps/instructor_task/tests/test_tasks_helper.py:
```diff
@@ -196,12 +198,62 @@ def test_report_stores_results(self, mock_time):
 
                 with patch('instructor_task.models.LocalFSReportStore.store_rows') as mock_store_rows:
-                    return_val = push_ora2_responses_to_s3(None, None, self.course.id, None, 'generated')
+
```
Nit: I don't think we want this blank line.

dylanrhodes commented on lms/djangoapps/instructor/views/instructor_dashboard.py:
```diff
@@ -330,6 +330,7 @@ def _section_data_download(course, access):
         'get_student_responses_url': reverse('get_student_responses', kwargs={'course_id': course_key.to_deprecated_string()}),
         'get_student_forums_usage_url': reverse('get_student_forums_usage', kwargs={'course_id': unicode(course_key)}),
         'get_ora2_responses_url': reverse('get_ora2_responses', kwargs={'course_id': course_key.to_deprecated_string()}),
+        'get_ora2_responses_url_email': reverse('get_ora2_responses_email', kwargs={'course_id': course_key.to_deprecated_string()}),
```
It seems like we could avoid a lot of duplicated code if instead of creating a new endpoint, we parameterized the existing one with an include_email boolean.  Have you thought about doing that instead?  I may not be correctly understanding the extra information included in the email query, but it looks very similar to the original one.  I would much rather parameterize and reuse the existing ORA2 report generation methods than duplicate them.

dylanrhodes commented on lms/djangoapps/instructor_task/tasks.py:
```diff
@@ -191,6 +193,7 @@ def get_course_forums_usage(entry_id, xmodule_instance_args):
     return run_main_task(entry_id, task_fn, action_name)
 
 
+# pylint: disable=not-callable
```
Why do we need to disable pylint for this existing method?  Is it raising linter exceptions?

dylanrhodes commented on lms/djangoapps/instructor/utils.py:
```diff
@@ -278,7 +292,7 @@ def ora2_data_queries():
     """
 
-    return RAW_QUERY
+    return RAW_QUERY_1 + RAW_QUERY_3 + RAW_QUERY_2
```
If it proves impossible to restructure the query above into a single parameterized string, we should probably still reorder/rename these variables to correspond to their order in the query.  I'd like this code to be crystal clear to people, so we don't have to worry about SQL attacks/misuse.

dylanrhodes commented on lms/djangoapps/instructor/utils.py:
```diff
@@ -185,16 +183,32 @@ def collect_ora2_data(course_id):
     Wraps a raw SQL query which retrieves all ORA2 responses for a course.
     """
-
-    RAW_QUERY = """
```
I would really prefer if we could maintain the existing structure of a single, parameterized SQL query here.  To my knowledge, this is the only section of the codebase in edx-platform which makes use of raw SQL querying, which I think is kind of dangerous, both from a SQL injection and possibility of people modifying it in the future without paying close attention.  At the very least, I think we can wrap RAW_QUERY_1 and RAW_QUERY_2 into a single string parameterized by RAW_QUERY_3 and course_id, but if possible, I'd like to keep all of the SQL in a single variable.

lydiali replied:
Changed this back to a single, parameterized SQL query.

dylanrhodes commented on lms/djangoapps/instructor/utils.py:
```diff
@@ -185,16 +183,32 @@ def collect_ora2_data(course_id):
-`student`.`student_id` AS `anonymized_student_id`,
+"""
+    RAW_QUERY_2 = """
+
```
Nit: I don't think we want this empty line.

dylanrhodes commented on lms/djangoapps/instructor/utils.py:
```diff
@@ -185,16 +183,32 @@ def collect_ora2_data(course_id):
-    RAW_QUERY = """
+    if is_anonymous:
+        RAW_QUERY_3 = """
+        `student`.`student_id` AS `anonymized_student_id`,
```
Nit: This doesn't have to be triple-quoted or split across lines, since it's a single-line string.

dylanrhodes commented on lms/djangoapps/instructor/utils.py:
```diff
@@ -185,16 +183,32 @@ def collect_ora2_data(course_id):
 SELECT `sub`.`uuid` AS `submission_uuid`,
 `student`.`item_id` AS `item_id`,
-`student`.`student_id` AS `anonymized_student_id`,
+"""
```
Nit: If this line stays in the method, the triple quotes should be aligned with RAW_QUERY_1.

azimpradhan commented on lms/templates/instructor/instructor_dashboard_2/data_download.html:
```diff
@@ -43,7 +43,8 @@
       <input type="button" name="get-student-responses" value="${_("Get Student Responses Report")}" data-endpoint="${ section_data['get_student_responses_url'] }"/>
-      <input type="button" name="ora2-response-btn" value="${_("Download Ora2 Responses")}" data-endpoint="${ section_data['get_ora2_responses_url'] }"/>
+      <input type="button" name="ora2-response-btn" value="${_("Download Ora2 Responses (no email addresses)")}" data-endpoint="${ section_data['get_ora2_responses_url'] }"/>
+      <input type="button" name="ora2-response-btn-email" value="${_("Download Ora2 Responses (including email addresses)")}" data-endpoint="${ section_data['get_ora2_responses_url_email'] }"/>
```
as far as naming convention, would it make more sense to use

get_ora2_email_responses_url  instead?

This would be more consistent with how other endpoints are named

azimpradhan replied:
however, if you are planning to use the same url (only parametrized with the anonymous boolean)  as @dylanrhodes suggested above, then you would use the same endpoint

azimpradhan commented on lms/djangoapps/instructor_task/tests/test_api.py:
```diff
@@ -220,6 +221,15 @@ def test_submit_ora2_request_task(self):
 
             mock_submit_task.assert_called_once_with(request, 'ora2_responses', get_ora2_responses, self.course.id, {}, '')
 
+    def test_submit_ora2_request_task_email(self):
```
similar to the comment below we could rename this to 

test_submit_ora2_email_request_task

azimpradhan commented on lms/djangoapps/instructor_task/tests/test_api.py:
```diff
@@ -24,7 +25,7 @@
 from instructor_task.api_helper import AlreadyRunningError
 from instructor_task.models import InstructorTask, PROGRESS
-from instructor_task.tasks import get_ora2_responses, get_course_forums_usage, get_student_forums_usage
+from instructor_task.tasks import get_ora2_responses, get_ora2_responses_email, get_course_forums_usage, get_student_forums_usage
```
get_ora2_email_responses

azimpradhan commented on lms/djangoapps/instructor/tests/test_api.py:
```diff
@@ -2545,6 +2546,24 @@ def test_get_student_forums_usage_already_running(self):
         already_running_status = "A student forums usage report task is already in progress."
         self.assertIn(already_running_status, response.content)
 
+    def test_get_ora2_responses_email_success(self):
```
if we perform renames, this would be test_get_ora2_email_reponses_success

azimpradhan commented:
I have finished reviewing this PR.

azimpradhan commented:
Key themes:
1) explore using a parameterized url as dylan suggested 
2) remove extra empty lines where pointed out
3) eliminate using the s3_base online helper method by directly calling _push_csv_responses_to_s3
4) choose a naming convention for methods that is consistent with the rest of the code

lydiali commented:
Thanks! I think I've incorporated all the suggestions outlined in Azim's key themes. Please review again. 
It's using a parameterized url as Dylan suggested, and as a result many of the duplicate methods were eliminated.

dcadams commented:
I've not been following this PR in detail but I believe there's a management command that uses the same code and wondered if that would be affected.

lydiali commented:
I tested the management command to make sure my changes didn't break that command.  I had it default to anonymous so we can keep the original behavior. (I can add a parameter if there is a need. I'm not clear on who runs this command) 

To run the management command, do this:
./manage.py lms collect_ora2_data <course id>  --settings=devstack

dylanrhodes commented on lms/djangoapps/instructor_task/tasks.py:
```diff
@@ -173,7 +173,7 @@ def get_student_responses(entry_id, xmodule_instance_args):
 def get_ora2_responses(entry_id, xmodule_instance_args):
     """
-    Generate a CSV of ora2 responses and push it to S3.
+    Generate a CSV of ora2 responses with anonymous id,  and push it to S3.
```
Now that the method is parameterized to handle both anonymous id reports and reports with emails attached, is this comment still true?  I might be missing something, but I think both types of reports hit this method.

dylanrhodes commented on lms/djangoapps/instructor_task/tasks.py:
```diff
@@ -117,7 +117,7 @@ def delete_problem_state(entry_id, xmodule_instance_args):
 @task(base=BaseInstructorTask, routing_key=settings.BULK_EMAIL_ROUTING_KEY)  # pylint: disable=E1102
 def send_bulk_course_email(entry_id, _xmodule_instance_args):
-    """Sends emails to recipients enrolled in a course.
+    """Sends emails to recipients ed in a course.
```
Nit: this looks like a minor typo.

dylanrhodes commented on lms/djangoapps/instructor/utils.py:
```diff
@@ -275,10 +272,22 @@ def ora2_data_queries():
     """
-
-    return RAW_QUERY
+    if include_email == 'False':
```
I'd suggest swapping the order of this conditional.  I think that positive conditionals are easier to read than negative ones.  It will also put the query section you want to use to include email addresses right next to the check on include_email's value.

dylanrhodes commented on lms/djangoapps/instructor/utils.py:
```diff
@@ -185,16 +183,15 @@ def collect_ora2_data(course_id):
 SELECT `sub`.`uuid` AS `submission_uuid`,
 `student`.`item_id` AS `item_id`,
-`student`.`student_id` AS `anonymized_student_id`,
+{0},
```
I'd suggest swapping this {0} for a keyword argument substitution like {id_column}.  Then your format call will look like "sql string".format(id_column=id_column), which I think is a little more explicit.  I'm not trying to be nitpicky; I just think query parameterization ought to be as clear as possible.

dylanrhodes commented on lms/djangoapps/instructor/utils.py:
```diff
@@ -185,16 +183,15 @@ def collect_ora2_data(course_id):
     """
     Wraps a raw SQL query which retrieves all ORA2 responses for a course.
     """
-
```
Nit: I think we should keep this blank line to separate the multiline comment from the multiline string constant.

dylanrhodes commented on lms/djangoapps/instructor/utils.py:
```diff
@@ -164,20 +164,18 @@ def generate_course_forums_query(course_id, query_type, parent_id_check=None):
-    
+    raw_queries = ora2_data_queries(include_email, course_id).split(';')
     cursor.execute(raw_queries[0])
-    cursor.execute(raw_queries[1], [course_id])
```
I would actually prefer to keep this execute line as is.  It's always better to use SQL libraries directly for query parameterization rather than other string substitution methods (like .format below).  Otherwise, you are almost guaranteed to leave yourself open to some form of injection attack, because SQL escaping is really finicky and the generic Python string substitution methods won't handle all of the edge cases correctly (<a href="http://xkcd.com/327/">my favorite XKCD</a>).

dylanrhodes commented on lms/djangoapps/instructor/utils.py:
```diff
@@ -164,20 +164,18 @@ def generate_course_forums_query(course_id, query_type, parent_id_check=None):
+def collect_ora2_data(course_id, include_email='False'):
     """
-    Query MySQL database for aggregated ora2 response data.
+    Query MySQL database for aggregated ora2 response data. If is_anonymous = false, display student username and emails.
```
I think this comment should reference include_email instead of is_anonymous.

dylanrhodes commented on lms/djangoapps/instructor_task/tasks_helper.py:
```diff
@@ -668,7 +660,7 @@ def push_student_forums_data_to_s3(_xmodule_instance_args, _entry_id, course_id,
 
 
-def _push_csv_responses_to_s3(csv_fn, filename, course_id, action_name):
+def _push_csv_responses_to_s3(csv_fn, filename, course_id, action_name, include_email=None):
```
Is there any way we can avoid changing this method's header to include include_email?  Lots of other reports use _push_csv_responses_to_s3, so changing its input arguments underneath them is not a great option.  I would almost rather separate out a collect_anonymous_ora2_data and collect_email_ora2_data method so that the include_email logic could be moved out of this method and included with the existing conditional in push_ora2_responses_to_s3.  Another option would be to change include_email to a more general parameter like kwargs with the same None default value.  Then, you could check to see if kwargs is None instead of include_email and pass it to the csv_fn if not.  This also would make the parameter available if others want to parameterize their data collection methods in the future.  I'm not enthusiastic about adding this parameter, though.

dylanrhodes commented:
Hey @lydiali this looks a lot better with the URL parameterization - thanks for adding that.  I have a few more suggestions, but I think it's getting pretty close to being able to merge.

lydiali commented:
Thanks @dylanrhodes for the detailed review.  

- I added collect_anonymous_ora2_data and collect_email_ora2_data methods to move the include_email logic out of _push_csv_responses_to_s3 method 
- changed the execute line back to cursor.execute(raw_queries[1], [course_id])
- swapping {0} for a keyword argument substitution like {id_column}.
- misc. white space

dylanrhodes commented on lms/djangoapps/instructor/utils.py:
```diff
@@ -164,28 +164,39 @@ def generate_course_forums_query(course_id, query_type, parent_id_check=None):
     """
-    cursor = get_read_replica_cursor_if_available(db)
+    include_email = 'False'
+    return collect_ora2_data(course_id, include_email)
```
I'd suggest swapping out these string values for actual booleans and also condensing the statement to a single line, so `return collect_ora2_data(course_id, False)` rather than this.  I also think the method comment should have more of an indication that it's a wrapper method, since this method doesn't really do the querying.  Maybe "Call collect_ora2_data for anonymized, aggregated ORA2 response data."?  (And below "Call collect_ora2_data for aggregated ORA2 response data including users' email addresses")

dylanrhodes commented:
I have two quick suggestions for the two new wrapper methods, but other than that, LGTM.  :+1: I won't be back until spring quarter, so maybe someone else can give it a once-over if we want to get it out the door in the meantime.

lydiali commented:
Please review again. thanks!

dcadams commented:
:+1:  
It's a pity this is the only place in the code that uses SQL but that's outside the scope of this PR.
